File: portal/src/lib/vnpay.ts
import { createHmac, timingSafeEqual } from 'node:crypto'
import type { Prisma } from '@prisma/client'

import prisma from '@/lib/prisma'

export type VNPayCreatePaymentResult = {
  paymentUrl: string
  orderId: string
}

type VNPayConfig = {
  tmnCode: string
  hashSecret: string
  paymentUrl: string
  amount: number
}

const DEFAULT_PRO_AMOUNT = 99000

function env(name: string, fallback = ''): string {
  return process.env[name]?.trim() || fallback
}

export function getVNPayConfig(): VNPayConfig | null {
  const tmnCode = env('VNPAY_TMN_CODE')
  const hashSecret = env('VNPAY_HASH_SECRET')
  const paymentUrl = env('VNPAY_PAYMENT_URL')
  if (!tmnCode || !hashSecret || !paymentUrl) {
    return null
  }

  const amount = Number.parseInt(env('VNPAY_PRO_AMOUNT', String(DEFAULT_PRO_AMOUNT)), 10)

  return {
    tmnCode,
    hashSecret,
    paymentUrl,
    amount: Number.isFinite(amount) && amount > 0 ? amount : DEFAULT_PRO_AMOUNT,
  }
}

function hmacSha512(data: string, secret: string): string {
  return createHmac('sha512', secret).update(Buffer.from(data, 'utf-8')).digest('hex')
}

function formatDate(date: Date) {
  const pad = (n: number) => String(n).padStart(2, '0')
  return [
    date.getFullYear(),
    pad(date.getMonth() + 1),
    pad(date.getDate()),
    pad(date.getHours()),
    pad(date.getMinutes()),
    pad(date.getSeconds()),
  ].join('')
}

function sortObject(input: Record<string, string>) {
  return Object.keys(input)
    .sort()
    .reduce<Record<string, string>>((acc, key) => {
      acc[key] = input[key]
      return acc
    }, {})
}

function buildQuery(data: Record<string, string>) {
  return new URLSearchParams(data).toString()
}

export function getVNPayProAmount(): number {
  return getVNPayConfig()?.amount ?? DEFAULT_PRO_AMOUNT
}

export function createVNPayPayment(args: {
  txnRef: string
  orderInfo: string
  returnUrl: string
  ipAddr: string
}) {
  const config = getVNPayConfig()
  if (!config) {
    throw new Error('VNPAY is not configured on the server.')
  }

  const requestData = sortObject({
    vnp_Version: '2.1.0',
    vnp_Command: 'pay',
    vnp_TmnCode: config.tmnCode,
    vnp_Amount: String(config.amount * 100),
    vnp_CurrCode: 'VND',
    vnp_TxnRef: args.txnRef,
    vnp_OrderInfo: args.orderInfo,
    vnp_OrderType: 'other',
    vnp_Locale: 'vn',
    vnp_ReturnUrl: args.returnUrl,
    vnp_IpAddr: args.ipAddr,
    vnp_CreateDate: formatDate(new Date()),
  })

  const signData = buildQuery(requestData)
  const secureHash = hmacSha512(signData, config.hashSecret)
  const paymentParams = new URLSearchParams({
    ...requestData,
    vnp_SecureHash: secureHash,
  })

  return {
    paymentUrl: `${config.paymentUrl}?${paymentParams.toString()}`,
    orderId: args.txnRef,
  } satisfies VNPayCreatePaymentResult
}

export function verifyVNPayPayload(rawPayload: URLSearchParams | Record<string, string>) {
  const config = getVNPayConfig()
  if (!config) return false

  const entries = rawPayload instanceof URLSearchParams ? Object.fromEntries(rawPayload.entries()) : rawPayload
  const provided = entries.vnp_SecureHash?.trim()
  if (!provided) return false

  const filtered: Record<string, string> = {}
  for (const [key, value] of Object.entries(entries)) {
    if (!key.startsWith('vnp_')) continue
    if (key === 'vnp_SecureHash' || key === 'vnp_SecureHashType') continue
    filtered[key] = value
  }

  const sorted = sortObject(filtered)
  const signData = buildQuery(sorted)
  const expected = hmacSha512(signData, config.hashSecret)
  const a = Buffer.from(provided)
  const b = Buffer.from(expected)
  if (a.length != b.length) return false
  return timingSafeEqual(a, b)
}

export async function processVNPayResult(rawPayload: URLSearchParams | Record<string, string>) {
  const entries = rawPayload instanceof URLSearchParams ? Object.fromEntries(rawPayload.entries()) : rawPayload
  const orderId = entries.vnp_TxnRef ?? ''
  const responseCode = entries.vnp_ResponseCode ?? ''
  const transactionStatus = entries.vnp_TransactionStatus ?? ''
  const amount = Number(entries.vnp_Amount ?? '0') / 100

  const transaction = await prisma.paymentTransaction.findUnique({
    where: { orderId },
  })

  if (!transaction) {
    return { ok: false as const, code: '01', message: 'Order not found' }
  }

  if (transaction.amount !== amount) {
    return { ok: false as const, code: '04', message: 'Invalid amount' }
  }

  if (transaction.status === 'paid') {
    return { ok: true as const, code: '02', message: 'Order already confirmed', success: true }
  }

  const isSuccess = responseCode === '00' && transactionStatus === '00'

  await prisma.paymentTransaction.update({
    where: { id: transaction.id },
    data: {
      status: isSuccess ? 'paid' : 'failed',
      providerRef: entries.vnp_TransactionNo ?? null,
      paidAt: isSuccess ? new Date() : null,
      rawNotification: entries as unknown as Prisma.InputJsonValue,
    },
  })

  if (isSuccess) {
    await prisma.user.update({
      where: { id: transaction.userId },
      data: { isPro: true },
    })
  }

  return {
    ok: true as const,
    code: '00',
    message: isSuccess ? 'Confirm Success' : 'Payment Failed',
    success: isSuccess,
  }
}

File: portal/api/internal/payments/vnpay.go
package payments

import (
	"context"
	"crypto/hmac"
	"crypto/sha512"
	"crypto/subtle"
	"encoding/hex"
	"encoding/json"
	"errors"
	"fmt"
	"net/url"
	"sort"
	"strconv"
	"strings"
	"time"

	"github.com/jackc/pgx/v5"
	"github.com/jackc/pgx/v5/pgxpool"
)

type VNPayConfig struct {
	TmnCode    string
	HashSecret string
	PaymentURL string
	Amount     int
}

type CreatePaymentArgs struct {
	TxnRef    string
	OrderInfo string
	ReturnURL string
	IPAddr    string
}

type CreatePaymentResult struct {
	PaymentURL string
	OrderID    string
}

type ProcessResult struct {
	OK      bool
	Code    string
	Message string
	Success bool
}

// CreateVNPayPayment generates a signed VNPay payment URL.
func CreateVNPayPayment(args CreatePaymentArgs, cfg VNPayConfig) (CreatePaymentResult, error) {
	if cfg.TmnCode == "" || cfg.HashSecret == "" || cfg.PaymentURL == "" {
		return CreatePaymentResult{}, errors.New("VNPAY is not configured on the server")
	}

	createDate := time.Now().Format("20060102150405")
	amountStr := strconv.Itoa(cfg.Amount * 100)

	params := url.Values{}
	params.Set("vnp_Version", "2.1.0")
	params.Set("vnp_Command", "pay")
	params.Set("vnp_TmnCode", cfg.TmnCode)
	params.Set("vnp_Amount", amountStr)
	params.Set("vnp_CurrCode", "VND")
	params.Set("vnp_TxnRef", args.TxnRef)
	params.Set("vnp_OrderInfo", args.OrderInfo)
	params.Set("vnp_OrderType", "other")
	params.Set("vnp_Locale", "vn")
	params.Set("vnp_ReturnUrl", args.ReturnURL)
	params.Set("vnp_IpAddr", args.IPAddr)
	params.Set("vnp_CreateDate", createDate)

	// Sort keys
	var keys []string
	for k := range params {
		keys = append(keys, k)
	}
	sort.Strings(keys)

	// Build query string matching URLSearchParams in Next.js.
	// URLSearchParams.toString() encodes spaces as "+" which is standard query string format.
	var queryParts []string
	for _, k := range keys {
		val := params.Get(k)
		if val != "" {
			queryParts = append(queryParts, url.QueryEscape(k)+"="+url.QueryEscape(val))
		}
	}
	signData := strings.Join(queryParts, "&")

	// Calculate secure hash
	mac := hmac.New(sha512.New, []byte(cfg.HashSecret))
	mac.Write([]byte(signData))
	secureHash := hex.EncodeToString(mac.Sum(nil))

	params.Set("vnp_SecureHash", secureHash)

	// Construct payment URL
	var finalParts []string
	for _, k := range keys {
		val := params.Get(k)
		if val != "" {
			finalParts = append(finalParts, url.QueryEscape(k)+"="+url.QueryEscape(val))
		}
	}
	finalParts = append(finalParts, "vnp_SecureHash="+url.QueryEscape(secureHash))

	paymentUrl := fmt.Sprintf("%s?%s", cfg.PaymentURL, strings.Join(finalParts, "&"))

	return CreatePaymentResult{
		PaymentURL: paymentUrl,
		OrderID:    args.TxnRef,
	}, nil
}

// VerifyVNPayPayload verifies the signature of a VNPay callback request in a timing-safe manner.
func VerifyVNPayPayload(params url.Values, hashSecret string) bool {
	if hashSecret == "" {
		return false
	}

	provided := strings.TrimSpace(params.Get("vnp_SecureHash"))
	if provided == "" {
		return false
	}

	// Filter out non-vnp parameters or signature parameters
	var keys []string
	for k := range params {
		if strings.HasPrefix(k, "vnp_") && k != "vnp_SecureHash" && k != "vnp_SecureHashType" {
			keys = append(keys, k)
		}
	}
	sort.Strings(keys)

	var queryParts []string
	for _, k := range keys {
		val := params.Get(k)
		if val != "" {
			queryParts = append(queryParts, url.QueryEscape(k)+"="+url.QueryEscape(val))
		}
	}
	signData := strings.Join(queryParts, "&")

	mac := hmac.New(sha512.New, []byte(hashSecret))
	mac.Write([]byte(signData))
	expected := hex.EncodeToString(mac.Sum(nil))

	a := []byte(provided)
	b := []byte(expected)
	if len(a) != len(b) {
		return false
	}

	return subtle.ConstantTimeCompare(a, b) == 1
}

// ProcessVNPayResult handles the transactional processing of the webhook callback.
func ProcessVNPayResult(ctx context.Context, dbPool dbPoolWrapper, params url.Values) (ProcessResult, error) {
	orderID := params.Get("vnp_TxnRef")
	responseCode := params.Get("vnp_ResponseCode")
	transactionStatus := params.Get("vnp_TransactionStatus")
	amountRaw := params.Get("vnp_Amount")

	var amountCent int
	if amountRaw != "" {
		if val, err := strconv.Atoi(amountRaw); err == nil {
			amountCent = val
		}
	}
	amount := amountCent / 100

	tx, err := dbPool.Begin(ctx)
	if err != nil {
		return ProcessResult{}, err
	}
	defer tx.Rollback(ctx)

	// Fetch transaction
	var txID string
	var txAmount int
	var txStatus string
	var txUserID *string
	var guestEmail *string

	err = tx.QueryRow(ctx, `
		SELECT "id", "amount", "status", "userId", "guestEmail"
		FROM "PaymentTransaction"
		WHERE "orderId" = $1
	`, orderID).Scan(&txID, &txAmount, &txStatus, &txUserID, &guestEmail)

	if err == pgx.ErrNoRows {
		return ProcessResult{OK: false, Code: "01", Message: "Order not found"}, nil
	} else if err != nil {
		return ProcessResult{}, err
	}

	if txAmount != amount {
		return ProcessResult{OK: false, Code: "04", Message: "Invalid amount"}, nil
	}

	if txStatus == "paid" {
		return ProcessResult{OK: true, Code: "02", Message: "Order already confirmed", Success: true}, nil
	}

	isSuccess := responseCode == "00" && transactionStatus == "00"

	// Marshal notifications params as json
	notificationMap := make(map[string]string)
	for k := range params {
		notificationMap[k] = params.Get(k)
	}
	notificationJSON, err := json.Marshal(notificationMap)
	if err != nil {
		return ProcessResult{}, err
	}

	status := "failed"
	if isSuccess {
		status = "paid"
	}

	var paidAt *time.Time
	if isSuccess {
		now := time.Now()
		paidAt = &now
	}

	providerRef := params.Get("vnp_TransactionNo")
	var providerRefPtr *string
	if providerRef != "" {
		providerRefPtr = &providerRef
	}

	_, err = tx.Exec(ctx, `
		UPDATE "PaymentTransaction"
		SET "status" = $1, "providerRef" = $2, "paidAt" = $3, "rawNotification" = $4, "updatedAt" = NOW()
		WHERE "id" = $5
	`, status, providerRefPtr, paidAt, notificationJSON, txID)
	if err != nil {
		return ProcessResult{}, err
	}

	if isSuccess {
		if txUserID != nil {
			_, err = tx.Exec(ctx, `
				UPDATE "User"
				SET "isPro" = true, "updatedAt" = NOW()
				WHERE "id" = $1
			`, *txUserID)
			if err != nil {
				return ProcessResult{}, err
			}
		} else if guestEmail != nil {
			// Check if guest has registered in the system
			var existingUserID string
			err = tx.QueryRow(ctx, `
				SELECT "id"
				FROM "User"
				WHERE LOWER("email") = LOWER($1)
			`, *guestEmail).Scan(&existingUserID)

			if err == nil {
				// Transition transaction and upgrade user
				_, err = tx.Exec(ctx, `
					UPDATE "PaymentTransaction"
					SET "userId" = $1, "updatedAt" = NOW()
					WHERE "id" = $2
				`, existingUserID, txID)
				if err != nil {
					return ProcessResult{}, err
				}

				_, err = tx.Exec(ctx, `
					UPDATE "User"
					SET "isPro" = true, "updatedAt" = NOW()
					WHERE "id" = $1
				`, existingUserID)
				if err != nil {
					return ProcessResult{}, err
				}
			} else if err != pgx.ErrNoRows {
				return ProcessResult{}, err
			}
		}
	}

	if err := tx.Commit(ctx); err != nil {
		return ProcessResult{}, err
	}

	message := "Payment Failed"
	if isSuccess {
		message = "Confirm Success"
	}

	return ProcessResult{
		OK:      true,
		Code:    "00",
		Message: message,
		Success: isSuccess,
	}, nil
}

// dbPoolWrapper bridges pgxpool.Pool interface so we can mock/run database calls
type dbPoolWrapper interface {
	Begin(ctx context.Context) (pgx.Tx, error)
	QueryRow(ctx context.Context, sql string, args ...any) pgx.Row
}

type RealDbPool struct {
	Pool *pgxpool.Pool
}

func (r *RealDbPool) Begin(ctx context.Context) (pgx.Tx, error) {
	return r.Pool.Begin(ctx)
}

func (r *RealDbPool) QueryRow(ctx context.Context, sql string, args ...any) pgx.Row {
	return r.Pool.QueryRow(ctx, sql, args...)
}

// WrapPool is a helper to wrap pgxpool
func WrapPool(pool *pgxpool.Pool) dbPoolWrapper {
	return &poolWrapper{pool: pool}
}

type poolWrapper struct {
	pool *pgxpool.Pool
}

func (pw *poolWrapper) Begin(ctx context.Context) (pgx.Tx, error) {
	return pw.pool.Begin(ctx)
}

func (pw *poolWrapper) QueryRow(ctx context.Context, sql string, args ...any) pgx.Row {
	return pw.pool.QueryRow(ctx, sql, args...)
}
